internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,255 @@
+	clientCancel context.CancelFunc
+
+	// tomb for the main loop
+	tomb *tomb.Tomb
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	g := &logGatherer{
+		targetName: target.Name,
+		tickPeriod: tickPeriod,
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+		tomb:       &tomb.Tomb{},
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	return g, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("plan changed")
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	if puller, ok := g.pullers[service.Name]; ok {
+		// This should never happen, but just in case, shut down the old puller.
+		_ = puller.tomb.Killf("service started")
+	}
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
+}
+
+func (g *logGatherer) loop() error {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-g.tomb.Dying():
+			return nil
+
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush(g.clientCtx)
+			if err != nil {
+				return fmt.Errorf("sending logs to target %q: %v", g.targetName, err)
+			}
+
+		case entry := <-g.entryCh:
+			err := g.client.Write(g.clientCtx, entry)
+			if err != nil {
+				return fmt.Errorf("writing logs to target %q: %v", g.targetName, err)
+			}
+		}
+	}
+}
+
+// stop closes the cancel channel, thereby terminating the main loop.
+// This method will block until gatherer teardown is complete.
+func (g *logGatherer) stop() {
+	// Start a timeout
+	time.AfterFunc(teardownTimeout, g.clientCancel)
+	// Kill the main loop
+	_ = g.tomb.Killf("gatherer closed")
+
+	// Shut down pullers
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("gatherer closed")
```

Comment:
Should the puller just have a reference to the gatherer's tomb, so that when the gatherer's tomb is killed, the pullers stop too, rather than having to do an explicit loop?

Reply:
I thought about this, but the pullers need a separate tomb because they sometimes need to stop independently of the gatherer. For example, if the plan changes, and the corresponding service is no longer forwarding logs to this target, we need to kill the puller's tomb even though the gatherer is not shutting down.